src/main/java/application/FormController.java

```diff
@@ -130,39 +128,53 @@ public SubmitResponseBean submitForm(@RequestBody SubmitRequestBean submitReques
                         purgeCasesTimer.durationInMs() > 2 ?
                                 "Puring cases took some time" : "Probably didn't have to purge cases");
-            } catch(InvalidStructureException e) {
-                submitResponseBean.setStatus(Constants.ANSWER_RESPONSE_STATUS_NEGATIVE);
-                submitResponseBean.setNotification(new NotificationMessage(e.getMessage(), true));
-                log.error("Submission failed with structure exception " + e);
-                return submitResponseBean;
-            }
-            ResponseEntity<String> submitResponse;
-            CategoryTimingHelper.RecordingTimer timer = categoryTimingHelper.newTimer(Constants.TimingCategories.SUBMIT_FORM_TO_HQ);
-            timer.start();
-            try {
-                 submitResponse = submitService.submitForm(
+
+                CategoryTimingHelper.RecordingTimer timer = categoryTimingHelper.newTimer(
+                        Constants.TimingCategories.SUBMIT_FORM_TO_HQ
+                );
+
+                timer.start();
+                ResponseEntity<String> submitResponse = submitService.submitForm(
                         formEntrySession.getInstanceXml(),
                         formEntrySession.getPostUrl()
                 );
-            } finally {
                 timer.end().record();
-            }
 
-            if (!submitResponse.getStatusCode().is2xxSuccessful()) {
-                submitResponseBean.setStatus("error");
-                submitResponseBean.setNotification(new NotificationMessage(
-                        "Form submission failed with error response" + submitResponse, true));
-                log.info("Submit response bean: " + submitResponseBean);
+                if (!submitResponse.getStatusCode().is2xxSuccessful()) {
+                    restoreFactory.rollback();
+                    submitResponseBean.setStatus("error");
+                    submitResponseBean.setNotification(new NotificationMessage(
+                            "Form submission failed with error response" + submitResponse, true));
+                    log.error("Submit response bean: " + submitResponseBean);
+                    return submitResponseBean;
+                }
+                // Only delete session immediately after successful submit
+                deleteSession(submitRequestBean.getSessionId());
+                restoreFactory.commit();
+
+            }
+            catch (InvalidStructureException e) {
+                restoreFactory.rollback();
+                submitResponseBean.setStatus(Constants.ANSWER_RESPONSE_STATUS_NEGATIVE);
+                submitResponseBean.setNotification(new NotificationMessage(e.getMessage(), true));
+                log.error("Submission failed with structure exception " + e);
                 return submitResponseBean;
             }
+            finally {
```

Comment:
@wpride Currently there are only two rollback conditions: Bad responses from the server, and Invalid structure.

I _think_ it's the case that we should be ready to handle any other deviant behavior as well though, right?

The pattern that we follow in android (thanks to good library support) is actually

```
transaction()
try{
   transaction.Succeesful();
} [potential catch blocks] 
finally {
   transaction.close();
}
```

Where ```transaction.close``` _automatically_ rolls back the commit without needing any specific error handler to do ti for us. The transaction essentially starts 'dirty' and has to be marked clean by the `try{}` block in order for ending the transaction to be a commit instead of a rollback.

Here, we should either be more proactive about handling unexpected situations, or we should adopt a similar approach where a rollback is the assumption, rather than an exceptional output.